fact_extractor/plugins/unpacking/tenvis_pk2/code/tenvis_pk2.py

```diff
@@ -0,0 +1,174 @@
+     4 | 16 | char[16] hash
+    20 |  4 | uint32 payload size
+    total size: 24 bytes
+    """
+
+    size = 24
+
+    def __init__(self, fp: BinaryIO, offset: int):
+        fp.seek(offset)
+        section_header_data = fp.read(self.size)
+        self.offset = offset
+        self.type = section_header_data[:4].rstrip(b'\x00').decode('ascii', errors='replace')
+        self.hash = section_header_data[4:20].hex()
+        self.payload_size = struct.unpack('<I', section_header_data[20:24])[0]
+
+    def to_dict(self):
+        return self.__dict__
+
+
+class Pk2File:
+    """
+     0  |  4 | uint32 filename size
+     4  |  x | char[x] filename
+    4+x |  4 | uint32 data size
+    x+8 |  y | ?? data (XOR encrypted)
+    total size: 4 + x + 4 + y bytes (== section payload size)
+    """
+
+    _BLOCK_SIZE = 1024  # we read the file block-wise to save memory
+
+    def __init__(self, fp: BinaryIO, offset: int, size: int):
+        self._fp = fp
+        self.size = size
+        self._file_offset = offset
+        self._fp.seek(self._file_offset)
+        filename_size = struct.unpack('<I', self._fp.read(4)[:4])[0]
+        self.filename = self._fp.read(filename_size).rstrip(b'\x00').decode('ascii', errors='replace')
+        self.data_offset = self._file_offset + 4 + filename_size + 4
+        self.data_size = struct.unpack('<I', self._fp.read(4))[0]
+
+    def save(self, save_dir: Path):
+        output_path = save_dir / self.filename.lstrip('/')
+        output_path.parent.mkdir(exist_ok=True, parents=True)
+        self._fp.seek(self.data_offset)
+        remaining = self.data_size
+        with output_path.open('wb') as out_fp:
+            while remaining > 0:
+                chunk_size = min(self._BLOCK_SIZE, remaining)
+                data = self._fp.read(chunk_size)
+                if not data:
+                    break
+                output = _decrypt(data)
+                out_fp.write(output)
+                remaining -= chunk_size
+
+    def to_dict(self):
+        return {k: v for k, v in self.__dict__.items() if not k.startswith('_')}
+
+
+class Pk2Cmd:
+    """
+    0 | x | char[x] command (XOR encrypted)
+    total size: x bytes (== section payload size)
+    """
+
+    def __init__(self, fp: BinaryIO, offset: int, size: int):
+        fp.seek(offset)
+        self.command = _decrypt(fp.read(size)).rstrip(b'\x00').decode('ascii', errors='replace')
+        self.size = size
+
+    def __str__(self):
+        cmd = repr(self.command)
+        return f'CMD: {cmd}'
+
+
+def _decrypt(data: bytes) -> bytearray:
+    output = bytearray()
+    for index, char in enumerate(data):
+        output.append(char ^ XOR_KEY[index % KEY_LEN])
+    return output
+
+
+def unpack_function(file_path: str, tmp_dir: str):
+    input_path, output_dir = Path(file_path), Path(tmp_dir)
+    meta = {'sections': []}
+
+    if input_path.stat().st_size < Pk2FileHeader.size + Pk2SectionHeader.size:
+        meta['errors'] = ['file too small']
```

Comment:
There can always only be one error based on this data flow, no? Then we should call this `error`  (singular)

Reply:
There can be multiple errors in the lower part of the function (where `meta.setdefault('errors', []).append()` is used). I wanted to keep the metadata consistent (even if there can be only one error in this case).

Reply:
I don't see how. There is only one item given to the `append` and there is a break directly behind. In any case: Consistent with what?

Reply:
I see now what you mean -> fixed